Set only the body of methods when we are in generation

goldhoorn opened the pull request.

goldhoorn commented:
This PR superseeds #40. The Problem is that the body can be left unknown for the spec. Nevertheless to keep backward compatibility we allow the function call on the task but ignore the body if we are not during generation phase.

The other solution would be to remove the attribute extension from the RTT::CPP part and save the body inside the spec. (What would not make sense either).

doudou commented:
Hi. While I like the intent, I would appreciate it if you could make it more complete. As you notice yourself, it makes no sense to have body specification in the spec part. We should therefore get rid of it, that is:

 - add a deprecation warning to the body argument in hidden_operation (and check how plain operations work)
 - make #base_body and #body DSL setters (using dsl_attribute) in the operation part
 - force people to call them in register_for_generation

goldhoorn commented:
don't have the time (now), if someone else would like to take over ping me otherwise it's on my todo
@christian-rauch ?

goldhoorn commented:
ok had found a few mins *updated*

goldhoorn commented:
*updated pr* i added a second call to usercpase plugins which allows to set/modify already created properties. Otherwise operation bodys cannot be setted/modified before the method generation.

doudou commented on lib/orogen/spec/task_context.rb:
```diff
@@ -758,10 +758,13 @@ def operation(name)
                 op = operation(name)
                 op.hidden = true
-                op.base_body = body
+                if(body)
```
Please don't put () around if arguments ...

doudou commented on lib/orogen/spec/task_context.rb:
```diff
@@ -758,10 +758,13 @@ def operation(name)
                 op.hidden = true
-                op.base_body = body
+                if(body)
+                    OroGen.warn "body argument for hidden_operation '#{self.name}'.'#{name}' is deprecated, please set the body during generation phase"
```
Mention the early_... method ?

goldhoorn commented:
*updated*

doudou commented on lib/orogen/gen/tasks.rb:
```diff
@@ -175,6 +175,10 @@ def register_for_generation
         # Spec::Operation
         module OperationGeneration
+
+            dsl_attribute(:body) { |value| value.to_s }
```
Indentation's fucked up

goldhoorn commented:
no clue how the tab was sneaked in but *fixed*
